commit b759f4ddcafb6414056cf3e49f2ab12359101c2e
Author: Felix Fietkau <nbd@openwrt.org>
Date:   Sat Feb 23 18:40:57 2013 +0100

    mac80211: fix idle handling in monitor mode
    
    When the driver does not want a monitor mode VIF, no channel context is
    allocated for it. This causes ieee80211_recalc_idle to put the hardware
    into idle mode if only a monitor mode is active, breaking injection.
    
    Fix this by checking local->monitors in addition to active channel
    contexts.
    
    Signed-off-by: Felix Fietkau <nbd@openwrt.org>
    Signed-off-by: Johannes Berg <johannes.berg@intel.com>

diff --git a/net/mac80211/iface.c b/net/mac80211/iface.c
index 2c059e54e885..640afab304d7 100644
--- a/net/mac80211/iface.c
+++ b/net/mac80211/iface.c
@@ -107,7 +107,7 @@ void ieee80211_recalc_idle(struct ieee80211_local *local)
 
 	lockdep_assert_held(&local->mtx);
 
-	active = !list_empty(&local->chanctx_list);
+	active = !list_empty(&local->chanctx_list) || local->monitors;
 
 	if (!local->ops->remain_on_channel) {
 		list_for_each_entry(roc, &local->roc_list, list) {